### ISSUE TITLE
Map view for tables does not display

### ISSUE THREAD

**Issue by cliftonmcintosh** [CodeforNepal/nepalmap_federal#49]
From a profile, click "Show data", then select "Map" tab. Map view does not display correctly. "Table" and "Distribution" display correctly.

Example: 

https://beta.nepalmap.org/data/map/?table=POPULATION&primary_geo_id=local-4003&geo_ids=local|district-28

![table-view-bug](https://user-images.githubusercontent.com/3824492/51495466-15310980-1d82-11e9-93f1-313ee956d00d.png)

Should look something like this:
https://nepalmap.org/data/map/?table=DISABILITY_SEX&primary_geo_id=district-04&geo_ids=district|country-NP

![map-view-in-pre-federal](https://user-images.githubusercontent.com/3824492/51493074-ad76c080-1d79-11e9-976b-6262f2db4a90.png)

**Comment by ravinepal** [comment]
@pratimakandel - would be interested in looking at this?

**Comment by pratimakandel** [comment]
Yes, I will try to look into it. And will let you know if I have any
question.

**Comment by ravinepal** [comment]
hi @hbvj99 - thank you for interest in contributing to open source products. Would you be interested in submitting a pull request for this issue?

**Comment by hbvj99** [comment]
@ravinepal thank you for mentioning. I'm working on a fix will notify you soon.

**Comment by hbvj99** [comment]
@ravinepal thank you for mentioning. I'm working on a fix and will notify you soon.

**Comment by clifton-mcintosh** [comment]
Please let me know if you have any questions while getting your dev environment set up.

**Comment by cliftonmcintosh** [comment]
One thing that may be useful is to run the app and

- monitor the web browser console
- take a look at the debug level SQL log statements that are put out in the terminal

There is a JavaScript error when this view is accessed, and it should provide insight into what is going wrong. This will almost certainly be very useful. If you set a breakpoint at the line that fails, you will see that each of the items in the `geoFeatures` array is undefined. This may be because of a mismatch between the geodata files and the database names for things. (It also may be something else.)

You might compare the debug level SQL log statements for the federal version of the app with the statements from running the pre-federal version of the app to see if there is something different.

Here are some screenshots of a Chrome console showing JS errors. The second one was taken with a breakpoint set at the place where the code fails, and it shows undefined geoFeatures when I tried to load the "show map" option.

Console error
![console-error](https://user-images.githubusercontent.com/3824492/53687416-fff0b880-3cf9-11e9-999a-59e42590decd.png)

geoFeatures undefined -- see the list of "local-xxxxx" on the right-hand side of this screenshot
![geofeatures-undefined](https://user-images.githubusercontent.com/3824492/53687799-b5256f80-3cfe-11e9-970d-785620c5b723.png)

**Comment by openrijal** [comment]
I was looking into this, had a few issues

1. Installation of the repo was difficult for me, some version mismatch involving Django, Wazimap and django-cors-header, had to install different versions to get the app up and running
2. Had to add auth module after upgrading to Django 2
3. The SQL that is run when the URL is hit, has these in it
`SELECT "wazimap_geography"."id", "wazimap_geography"."geo_level", "wazimap_geography"."geo_code", "wazimap_geography"."version", "wazimap_geography"."name", "wazimap_geography"."long_name", "wazimap_geography"."square_kms", "wazimap_geography"."parent_level", "wazimap_geography"."parent_code" FROM "wazimap_geography" WHERE ("wazimap_geography"."geo_code" = '28' AND "wazimap_geography"."geo_level" = 'district') ORDER BY "wazimap_geography"."version" DESC LIMIT 1; args=(u'28', u'district')
SELECT "wazimap_geography"."id", "wazimap_geography"."geo_level", "wazimap_geography"."geo_code", "wazimap_geography"."version", "wazimap_geography"."name", "wazimap_geography"."long_name", "wazimap_geography"."square_kms", "wazimap_geography"."parent_level", "wazimap_geography"."parent_code" FROM "wazimap_geography" WHERE ("wazimap_geography"."parent_level" = 'district' AND "wazimap_geography"."version" = '2016' AND "wazimap_geography"."parent_code" = '28'); args=(u'district', u'2016', u'28')
SELECT "wazimap_geography"."id", "wazimap_geography"."geo_level", "wazimap_geography"."geo_code", "wazimap_geography"."version", "wazimap_geography"."name", "wazimap_geography"."long_name", "wazimap_geography"."square_kms", "wazimap_geography"."parent_level", "wazimap_geography"."parent_code" FROM "wazimap_geography" WHERE ("wazimap_geography"."parent_level" = 'local' AND "wazimap_geography"."version" = '2016' AND "wazimap_geography"."parent_code" = '4001'); args=(u'local', u'2016', u'4001')
SELECT "wazimap_geography"."id", "wazimap_geography"."geo_level", "wazimap_geography"."geo_code", "wazimap_geography"."version", "wazimap_geography"."name", "wazimap_geography"."long_name", "wazimap_geography"."square_kms", "wazimap_geography"."parent_level", "wazimap_geography"."parent_code" FROM "wazimap_geography" WHERE ("wazimap_geography"."parent_level" = 'local' AND "wazimap_geography"."version" = '2016' AND "wazimap_geography"."parent_code" = '4002'); args=(u'local', u'2016', u'4002')
SELECT "wazimap_geography"."id", "wazimap_geography"."geo_level", "wazimap_geography"."geo_code", "wazimap_geography"."version", "wazimap_geography"."name", "wazimap_geography"."long_name", "wazimap_geography"."square_kms", "wazimap_geography"."parent_level", "wazimap_geography"."parent_code" FROM "wazimap_geography" WHERE ("wazimap_geography"."parent_level" = 'local' AND "wazimap_geography"."version" = '2016' AND "wazimap_geography"."parent_code" = '4003'); args=(u'local', u'2016', u'4003')
SELECT "wazimap_geography"."id", "wazimap_geography"."geo_level", "wazimap_geography"."geo_code", "wazimap_geography"."version", "wazimap_geography"."name", "wazimap_geography"."long_name", "wazimap_geography"."square_kms", "wazimap_geography"."parent_level", "wazimap_geography"."parent_code" FROM "wazimap_geography" WHERE ("wazimap_geography"."parent_level" = 'local' AND "wazimap_geography"."version" = '2016' AND "wazimap_geography"."parent_code" = '4004'); args=(u'local', u'2016', u'4004')
SELECT "wazimap_geography"."id", "wazimap_geography"."geo_level", "wazimap_geography"."geo_code", "wazimap_geography"."version", "wazimap_geography"."name", "wazimap_geography"."long_name", "wazimap_geography"."square_kms", "wazimap_geography"."parent_level", "wazimap_geography"."parent_code" FROM "wazimap_geography" WHERE ("wazimap_geography"."parent_level" = 'local' AND "wazimap_geography"."version" = '2016' AND "wazimap_geography"."parent_code" = '4005'); args=(u'local', u'2016', u'4005')
SELECT "wazimap_geography"."id", "wazimap_geography"."geo_level", "wazimap_geography"."geo_code", "wazimap_geography"."version", "wazimap_geography"."name", "wazimap_geography"."long_name", "wazimap_geography"."square_kms", "wazimap_geography"."parent_level", "wazimap_geography"."parent_code" FROM "wazimap_geography" WHERE ("wazimap_geography"."parent_level" = 'local' AND "wazimap_geography"."version" = '2016' AND "wazimap_geography"."parent_code" = '4006'); args=(u'local', u'2016', u'4006')
SELECT "wazimap_geography"."id", "wazimap_geography"."geo_level", "wazimap_geography"."geo_code", "wazimap_geography"."version", "wazimap_geography"."name", "wazimap_geography"."long_name", "wazimap_geography"."square_kms", "wazimap_geography"."parent_level", "wazimap_geography"."parent_code" FROM "wazimap_geography" WHERE ("wazimap_geography"."parent_level" = 'local' AND "wazimap_geography"."version" = '2016' AND "wazimap_geography"."parent_code" = '4007'); args=(u'local', u'2016', u'4007')
SELECT "wazimap_geography"."id", "wazimap_geography"."geo_level", "wazimap_geography"."geo_code", "wazimap_geography"."version", "wazimap_geography"."name", "wazimap_geography"."long_name", "wazimap_geography"."square_kms", "wazimap_geography"."parent_level", "wazimap_geography"."parent_code" FROM "wazimap_geography" WHERE ("wazimap_geography"."parent_level" = 'local' AND "wazimap_geography"."version" = '2016' AND "wazimap_geography"."parent_code" = '4008'); args=(u'local', u'2016', u'4008')
SELECT "wazimap_geography"."id", "wazimap_geography"."geo_level", "wazimap_geography"."geo_code", "wazimap_geography"."version", "wazimap_geography"."name", "wazimap_geography"."long_name", "wazimap_geography"."square_kms", "wazimap_geography"."parent_level", "wazimap_geography"."parent_code" FROM "wazimap_geography" WHERE ("wazimap_geography"."parent_level" = 'local' AND "wazimap_geography"."version" = '2016' AND "wazimap_geography"."parent_code" = '4009'); args=(u'local', u'2016', u'4009')
SELECT "wazimap_geography"."id", "wazimap_geography"."geo_level", "wazimap_geography"."geo_code", "wazimap_geography"."version", "wazimap_geography"."name", "wazimap_geography"."long_name", "wazimap_geography"."square_kms", "wazimap_geography"."parent_level", "wazimap_geography"."parent_code" FROM "wazimap_geography" WHERE ("wazimap_geography"."parent_level" = 'local' AND "wazimap_geography"."version" = '2016' AND "wazimap_geography"."parent_code" = '4010'); args=(u'local', u'2016', u'4010')
SELECT "wazimap_geography"."id", "wazimap_geography"."geo_level", "wazimap_geography"."geo_code", "wazimap_geography"."version", "wazimap_geography"."name", "wazimap_geography"."long_name", "wazimap_geography"."square_kms", "wazimap_geography"."parent_level", "wazimap_geography"."parent_code" FROM "wazimap_geography" WHERE ("wazimap_geography"."parent_level" = 'local' AND "wazimap_geography"."version" = '2016' AND "wazimap_geography"."parent_code" = '4011'); args=(u'local', u'2016', u'4011')
SELECT "wazimap_geography"."id", "wazimap_geography"."geo_level", "wazimap_geography"."geo_code", "wazimap_geography"."version", "wazimap_geography"."name", "wazimap_geography"."long_name", "wazimap_geography"."square_kms", "wazimap_geography"."parent_level", "wazimap_geography"."parent_code" FROM "wazimap_geography" WHERE ("wazimap_geography"."parent_level" = 'local' AND "wazimap_geography"."version" = '2016' AND "wazimap_geography"."parent_code" = '4012'); args=(u'local', u'2016', u'4012')
SELECT "wazimap_geography"."id", "wazimap_geography"."geo_level", "wazimap_geography"."geo_code", "wazimap_geography"."version", "wazimap_geography"."name", "wazimap_geography"."long_name", "wazimap_geography"."square_kms", "wazimap_geography"."parent_level", "wazimap_geography"."parent_code" FROM "wazimap_geography" WHERE ("wazimap_geography"."parent_level" = 'local' AND "wazimap_geography"."version" = '2016' AND "wazimap_geography"."parent_code" = '4013'); args=(u'local', u'2016', u'4013')
SELECT "wazimap_geography"."id", "wazimap_geography"."geo_level", "wazimap_geography"."geo_code", "wazimap_geography"."version", "wazimap_geography"."name", "wazimap_geography"."long_name", "wazimap_geography"."square_kms", "wazimap_geography"."parent_level", "wazimap_geography"."parent_code" FROM "wazimap_geography" WHERE ("wazimap_geography"."parent_level" = 'local' AND "wazimap_geography"."version" = '2016' AND "wazimap_geography"."parent_code" = '4014'); args=(u'local', u'2016', u'4014')
SELECT "wazimap_geography"."id", "wazimap_geography"."geo_level", "wazimap_geography"."geo_code", "wazimap_geography"."version", "wazimap_geography"."name", "wazimap_geography"."long_name", "wazimap_geography"."square_kms", "wazimap_geography"."parent_level", "wazimap_geography"."parent_code" FROM "wazimap_geography" WHERE ("wazimap_geography"."parent_level" = 'local' AND "wazimap_geography"."version" = '2016' AND "wazimap_geography"."parent_code" = '4015'); args=(u'local', u'2016', u'4015')`

The first 2 queries generate good data, the others don't, does this look right? @clifton-mcintosh

**Comment by cliftonmcintosh** [comment]
@crackjack

For the purposes of this ticket, I would suggest not doing any library upgrades so that those changes don't mask or confound the issue in this ticket.

I do not know what upgrading Django will do. It may or may not be compatible with wazimap 0.8.2. I would suggest creating a separate issue for that so that any changes that need to be made for possible Django upgrades be separated out from the issue reported here.

#30 covers upgrading the wazimap framework to 1.0. Even then, I am not sure about compatibility with Django 2.0. Please note that Wazimap South Africa uses Django 1.9.2 with Wazimap 1.1.1. https://github.com/OpenUpSA/wazimap-za/blob/master/requirements.txt

Based on the information that I see from the Chrome developer console, I am fairly certain that our problem in this issue is a mismatch between our geo files and how Wazimap fetches data from them.

**Comment by cliftonmcintosh** [comment]
@hbvj99 

You reached out to me by email to ask about this ticket a couple of weeks ago. Did my updated comments help provide any guidance on this?

**Comment by hbvj99** [comment]
@cliftonmcintosh It seems like there is a data mismatch between topojson, geojson. Also, when I replaced district topojson from nepalmap_app, the map seems to be working fine, but the geolocation is not correct.

![Screenshot from 2019-03-10 21-46-33](https://user-images.githubusercontent.com/43197293/54088128-953b1f00-4382-11e9-9666-8081f967ab48.png)

**Comment by cliftonmcintosh** [comment]
@hbvj99 

When the app goes into that view, it tries to find the matching area in the topojson files by using the "geoid" fields in the topojson files. It expects the "geoid" fields to be the "geo_level" plus the "geo_code", like "level-code". For example, if it is trying to find the geographic data about Achham, it will look in the "district.topojson" file for `"geoid":"district-01"`. If you look at the district.topojson file, you will see that the `geoid` fields in that file do not match what Wazimap needs. For example, Achham's "geoid" in that file is "dis-01". So the map view fails. The "geoid" fields in all our topojson files need to conform to the requirements of Wazimap. They need to be of the form "geo_level-geo_code".

**Comment by openrijal** [comment]
I'm actually editing the topojson and geojson files to make that change from `dis-01` to `district-01` and `loc-1001` to `local-1001`.

I found that was the only change from old nepalmap v/s the federal one in terms of geo codes

**Comment by openrijal** [comment]
for my local installation, I had to upgrade both wazimap and django to get it running, but I will make sure not to push those changes so that it does not break anything.

**Comment by clifton-mcintosh** [comment]
It sounds like we have two people working on the issue, @crackjack and @hbvj99

**Comment by openrijal** [comment]
in that case, I'll stop working on it. I have changed the geojson and topojson files, but looks like we also need to edit those entries in the database to match them.

I was thinking to change the SQL files next, but I'll wait until further notice. I am available to help if needed @hbvj99

**Comment by cliftonmcintosh** [comment]
I don't believe we need to change the SQL

**Comment by openrijal** [comment]
the sql file has `geo_codes` as just integers, is that fine? I thought since we said the `geo_codes` should be changed in the topojson and geojson, we need to change it in the sql too.

I think I should start reading about it in the wazimap documentation instead of asking you all these questions. @cliftonmcintosh 

any updates on this @hbvj99? lets sync up if you want to collaborate on this.

**Comment by cliftonmcintosh** [comment]
@crackjack 

Even though they don't have quotes around them in the SQL file, they are read in as strings. You can see the table definition here:
https://github.com/Code4Nepal/nepalmap_federal/blob/dev/sql/geography.sql#L34-L44

Note the definition of `geo_code`: `geo_code character varying(10) NOT NULL`.

FWIW, the `wazimap_geography` table structure is defined by Wazimap. If we altered the data types, it would cause the framework to fail.

**Comment by hbvj99** [comment]
@cliftonmcintosh @crackjack I'm not sure how to tackle this issue, do we have to update geo files? I found ```wazimap_za``` https://github.com/OpenUpSA/wazimap-za don't use geo files a static content. Could we apply the same approach on solving this issue?

**Comment by openrijal** [comment]
I have updated the geojson and topojson files with the geo_id being `local-`, `district-` etc, I can push it tonight if you want me to.

**Comment by cliftonmcintosh** [comment]
@crackjack 

I'm curious if you've had a chance to test the impact of the changes. Does the map view render correctly after the changes?

**Comment by cliftonmcintosh** [comment]
@hbvj99 

I believe Wazimap South Africa uses https://mapit.code4sa.org/ to manage its geographical lookups. See https://github.com/OpenUpSA/wazimap-za/blob/master/wazimap_za/geo.py

The standard way to manage map data in the Wazimap framework is to use geojson and topojson files. See https://wazimap.readthedocs.io/en/latest/geos.html#maps-and-boundary-data

**Comment by openrijal** [comment]
@cliftonmcintosh I tried to install the dependencies from our project, as well as from the wazimap_za repo's requirement, but I cannot get past the dependency issue
```
wazimap 2.0.0 has requirement Django<1.12.0,>=1.11.15, but you'll have django 1.9.2 which is incompatible.
django-cors-headers 2.5.2 has requirement Django>=1.11, but you'll have django 1.9.2 which is incompatible.
```

**Comment by cliftonmcintosh** [comment]
The requirements.txt for this project specify wazimap 0.8.2.

https://github.com/Code4Nepal/nepalmap_federal/blob/dev/requirements.txt

The printout from your message says wazimap 2.0.0. I'm not sure how that got there. Can you set up virtualenv or some other environment manager and run `pip install requirements.txt` so that it uses the correct version of wazimap? When I run `pip show wazimap`, it tells me I'm on 0.8.2. That's what is running on the servers, too.

We have an issue to upgrade to wazimap 1.0.0. https://github.com/Code4Nepal/nepalmap_federal/issues/30

**Comment by openrijal** [comment]
I'm sorry about not explaining it fully. Let me put it this way:
1. using the requirements from our project, it gets 0.8.2 for wazimap, no doubt. but I get an error on middlewares for django-cors-headers because it requires django version above 1.10
2. using the requirements from wazimap_za repo, it fails to compile GDAL, and I get wazimap 1.0.0.

I have used virtualenv, I do that for every project. If you can `pip freeze` your environment and send it over that would be very helpful. Also, I am using Pytyon 2.7, using Python 3 has a lot of other issues, which I believe is not what we want .

I had a successful combination of django, django-cors-headers and wazimap upgraded, that caused an error with JavaScript, and hence I was not able to get the map loading.

**Comment by openrijal** [comment]
I say we setup a time of about max 1 hour to go over this, it will help me catch up and may be fix the issue ultimately.

**Comment by cliftonmcintosh** [comment]
@crackjack 

Here is the outcome of `pip freeze`: [nepalmap_federal.txt](https://github.com/Code4Nepal/nepalmap_federal/files/2973929/nepalmap_federal.txt)

And here is my python version:
```
$ python --version                 
Python 2.7.15rc1
```
What OS are you running? I can spin up a VM with that OS and try to build the project fresh to see if I encounter the problems you have encountered.

I doubt I have any advice that is better conveyed verbally, but fee free to contact me directly via email if you would like to set something up.

**Comment by openrijal** [comment]
see you have different versions of Django, Gdal, etc

**Comment by openrijal** [comment]
I'll get this running with your requirement, I hope it will work. I use Arch Linux, so it doesn't have a version, may be using it through docker is a better idea.

**Comment by cliftonmcintosh** [comment]
@crackjack 

I did not try with Arch, but I was able to do an install in Linux Mint. The one package import I had to adjust was `django-cors-headers`, setting it to 2.4.2 in the requirements.txt. `pip install` loaded `django-cors-headers 2.5.2` by default and it failed on a deprecated dependency. Once I adjusted requirements.txt for that, I was fine.

**Comment by openrijal** [comment]
@cliftonmcintosh you were right about the dependencies. I re-did everything, installed from requirements.txt file but with version `django-cors-headers==2.4.1` because 2.4.2 was not found.

I am able to load the map now, I have attached a screenshot here.

I am going to push it and create a pull request for you to review, I have also changed the requirements.txt file with the change.

![Screenshot from 2019-03-16 20-45-22](https://user-images.githubusercontent.com/191368/54483627-dcb32680-482c-11e9-860a-c6cfa1d5d5b1.png)

**Comment by openrijal** [comment]
PR for review: https://github.com/Code4Nepal/nepalmap_federal/pull/60

@cliftonmcintosh @ravinepal

**Comment by cliftonmcintosh** [comment]
Fixed with #60 

Thank you, @crackjack

**Comment by cliftonmcintosh** [comment]
@crackjack 

Your changes are now deployed to https://beta.nepalmap.org

Thank you for the fix.
